File: PST/FUEL/src/FUEL/FS/Misc.vb
Imports DataMatrix.net
Imports FUEL
Imports FUEL.My
Imports MessagingToolkit.QRCode.Codec
Imports Microsoft.VisualBasic
Imports Microsoft.VisualBasic.CompilerServices
Imports System
Imports System.Diagnostics
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.Drawing.Imaging
Imports System.IO
Imports System.Reflection
Imports System.Text

Namespace FUEL.FS
    <StandardModule> _
    Public NotInheritable Class Misc
        ' Methods
        Private Shared Function AddTextToBarcode(ByVal bcdText As String, ByVal OriginalImage As Bitmap, ByVal TransParentBackground As Boolean) As Bitmap
            Dim brush As Brush = Nothing
            brush = If(TransParentBackground, Brushes.White, Brushes.Snow)
            Dim font As New Font("Courier New", 24!)
            Dim text As String = bcdText
            Dim ef As SizeF = Graphics.FromImage(OriginalImage).MeasureString([text], font)
            Dim image As New Bitmap(CInt(Math.Round(CDbl(((OriginalImage.Width + ef.Width) + 12!)))), (OriginalImage.Height + 12), PixelFormat.Format24bppRgb)
            Dim graphics2 As Graphics = Graphics.FromImage(image)
            Dim objA As Graphics = graphics2
            Try 
                graphics2.SmoothingMode = SmoothingMode.AntiAlias
                Dim rect As New RectangleF(-1!, -1!, CSng((image.Width + 2)), CSng((image.Height + 2)))
                graphics2.FillRectangle(brush, rect)
                Dim rectangle As New Rectangle(6, 6, image.Width, image.Height)
                graphics2.DrawImageUnscaled(OriginalImage, rectangle)
                rect = New RectangleF(CSng((OriginalImage.Width + 10)), 0!, CSng(image.Width), CSng(image.Height))
                graphics2.DrawString([text], font, Brushes.Red, rect)
            Finally
                If Not Object.ReferenceEquals(objA, Nothing) Then
                    objA.Dispose
                End If
            End Try
            Return image
        End Function

        Public Shared Sub CheckForDAQDrivers()
            Dim str As String = Path.Combine(New FileInfo(Assembly.GetExecutingAssembly.Location).DirectoryName, "LaunchNIDaqInstall.exe")
            Dim process As New Process
            process.StartInfo.WindowStyle = ProcessWindowStyle.Hidden
            process.StartInfo.FileName = str
            process.Start
            process.WaitForExit
        End Sub

        Public Shared Sub CheckForUpdate(ByVal UpdateType As CheckType)
            modCheckForUpdates.CheckForUpdate(CheckType.FS)
        End Sub

        Private Shared Function ConvertByteArrayToString(ByVal byteArray As Byte()) As String
            Return Encoding.UTF8.GetString(byteArray)
        End Function

        Public Shared Sub Create2dBarcode(ByVal PSID As String, ByVal PrinterSerialNum As String, ByVal OutputPath As String, ByVal TransParentBackground As Boolean)
            FileSystem.SetPath(Path.GetDirectoryName(OutputPath))
            Dim color As New Color
            color = If(TransParentBackground, Color.White, Color.Snow)
            Dim options As New DmtxImageEncoderOptions With { _
                .ModuleSize = 9, _
                .BackColor = color, _
                .MarginSize = 0, _
                .ForeColor = Color.Black _
            }
            Dim strArray As String() = New String() { PSID, ChrW(13) & ChrW(10), PrinterSerialNum, ChrW(13) & ChrW(10), Conversions.ToString(DateAndTime.Now), ChrW(13) & ChrW(10) & "FUEL: ", TestInformation.FuelRev }
            Misc.AddTextToBarcode(String.Concat(strArray), New DmtxImageEncoder().EncodeImage(PSID, options), TransParentBackground).Save(OutputPath, ImageFormat.Bmp)
        End Sub

        Public Shared Sub CreateBMPFromString(ByVal [Text] As String, ByVal OutputPath As String, ByVal TransParentBackground As Boolean)
            Dim brush As Brush = Nothing
            brush = If(TransParentBackground, Brushes.White, Brushes.Snow)
            Dim font As New Font("Arial", 20!)
            Dim ef As SizeF = Graphics.FromImage(New Bitmap(1, 1, PixelFormat.Format24bppRgb)).MeasureString([Text], font)
            Dim image As New Bitmap(CInt(Math.Round(CDbl(ef.Width))), CInt(Math.Round(CDbl(ef.Height))), PixelFormat.Format24bppRgb)
            Dim graphics As Graphics = Graphics.FromImage(image)
            Dim objA As Graphics = graphics
            Try 
                Dim rect As New Rectangle(-1, -1, CInt(Math.Round(CDbl((ef.Width + 2!)))), CInt(Math.Round(CDbl((ef.Height + 2!)))))
                graphics.FillRectangle(brush, rect)
                rect = New Rectangle(1, 1, CInt(Math.Round(CDbl((ef.Width + 2!)))), CInt(Math.Round(CDbl((ef.Height + 2!)))))
                graphics.DrawString([Text], font, Brushes.Red, rect)
            Finally
                If Not Object.ReferenceEquals(objA, Nothing) Then
                    objA.Dispose
                End If
            End Try
            image.Save(OutputPath)
        End Sub

        Public Shared Sub CreateQRCode(ByVal PSID As String, ByVal PrinterSerialNum As String, ByVal OutputPath As String)
            Misc.CreateQRCode(PSID, PrinterSerialNum, OutputPath, 4, 2, False)
        End Sub

        Public Shared Sub CreateQRCode(ByVal PSID As String, ByVal PrinterSerialNum As String, ByVal OutputPath As String, ByVal Scale As Integer, ByVal ErrorCorrection As Integer, ByVal TransParentBackground As Boolean)
            FileSystem.SetPath(Path.GetDirectoryName(OutputPath))
            Dim color As New Color
            color = If(TransParentBackground, Color.White, Color.Snow)
            Dim originalImage As Bitmap = New QRCodeEncoder() With { _
                .QRCodeErrorCorrect = DirectCast(ErrorCorrection, ERROR_CORRECTION), _
                .QRCodeScale = Scale, _
                .QRCodeEncodeMode = ENCODE_MODE.ALPHA_NUMERIC, _
                .QRCodeBackgroundColor = color _
            }.Encode(PSID)
            originalImage.SetPixel(1, 1, Color.Red)
            Dim strArray As String() = New String() { PSID, ChrW(13) & ChrW(10), PrinterSerialNum, ChrW(13) & ChrW(10), Conversions.ToString(DateAndTime.Now), ChrW(13) & ChrW(10) & "FUEL: ", TestInformation.FuelRev }
            Misc.AddTextToBarcode(String.Concat(strArray), originalImage, TransParentBackground).Save(OutputPath, ImageFormat.Bmp)
        End Sub

        Public Shared Function GenerateFUELCompatibleFileName(ByVal Product As String, ByVal DataType As String, ByVal ValidDuration As FileType, ByVal FileExtension As String) As Object
            Return Misc.GenerateFUELCompatibleFileName(Product, DataType, ValidDuration, FileExtension, Nothing)
        End Function

        Public Shared Function GenerateFUELCompatibleFileName(ByVal Product As String, ByVal DataType As String, ByVal ValidDuration As FileType, ByVal FileExtension As String, ByVal ExtraInfo As String) As Object
            Dim right As String = Strings.Replace(Conversions.ToString(DateAndTime.Now.Date), "/", "-", 1, -1, CompareMethod.Binary)
            If (ValidDuration = FileType.Monthly) Then
                right = Strings.Replace((Conversions.ToString(DateAndTime.Now.Month) & "-" & Conversions.ToString(DateAndTime.Now.Year)), "/", "-", 1, -1, CompareMethod.Binary)
            End If
            If Not FileExtension.StartsWith(".") Then
                FileExtension = ("." & FileExtension)
            End If
            Dim str As String = Conversions.ToString(Operators.ConcatenateObject(Operators.ConcatenateObject(Operators.ConcatenateObject(((DataType & "_") & Product & "_"), Misc.GetComputerName(True)), "_"), right))
            If (ExtraInfo <> Nothing) Then
                str = (str & "_" & ExtraInfo)
            End If
            Return (str & FileExtension)
        End Function

        Public Shared Function GetComputerName() As Object
            Return Misc.GetComputerName(False)
        End Function

        Friend Shared Function GetComputerName(ByVal ForFUELCompatFileName As Boolean) As Object
            Return If(ForFUELCompatFileName, MyProject.Computer.Name.Replace("_", "-"), MyProject.Computer.Name)
        End Function

        Public Shared Function GetMediaSize(ByVal File As String) As Integer
            FileSystem.SetPath(File, FilePurpose.ForReading)
            Dim array As Byte() = MyProject.Computer.FileSystem.ReadAllBytes(File)
            Dim flag As Boolean = False
            Dim strArray As String() = New String(1  - 1) {}
            Dim num6 As Long = Information.UBound(array, 1)
            Dim num3 As Long = 0
            Do While True
                Dim num8 As Long = num6
                If (num3 > num8) Then
                    Exit Do
                End If
                Dim flag3 As Boolean = (array(CInt(num3)) = &H26)
                If (flag3 AndAlso (array(CInt((num3 + 1))) = &H6C)) Then
                    Dim flag2 As Boolean = False
                    Dim num4 As Long = num3
                    Do While True
                        Dim buffer2 As Byte()
                        If Not (Not flag2 And (num4 < Information.UBound(array, 1))) Then
                            Exit Do
                        End If
                        If (strArray(0) = Nothing) Then
                            buffer2 = New Byte() { array(CInt(num4)) }
                            strArray(0) = Misc.ConvertByteArrayToString(buffer2)
                        Else
                            strArray = DirectCast(Utils.CopyArray(DirectCast(strArray, Array), New String(((Information.UBound(strArray, 1) + 1) + 1)  - 1) {}), String())
                            buffer2 = New Byte() { array(CInt(num4)) }
                            strArray(Information.UBound(strArray, 1)) = Misc.ConvertByteArrayToString(buffer2)
                        End If
                        buffer2 = New Byte() { array(CInt(num4)) }
                        If (Misc.ConvertByteArrayToString(buffer2).ToLower <> "a") Then
                            buffer2 = New Byte() { array(CInt(num4)) }
                            If (Misc.ConvertByteArrayToString(buffer2) = ChrW(27)) Then
                                flag2 = True
                                flag = False
                                strArray = New String(1  - 1) {}
                            End If
                            num4 = (num4 + 1)
                        Else
                            flag = True
                            flag2 = True
                            Exit Do
                        End If
                    Loop
                End If
                num3 = (num3 + 1)
            Loop
            Dim num2 As Integer = -100
            If flag Then
                Dim expression As String = Nothing
                Dim num7 As Integer = (Information.UBound(strArray, 1) - 1)
                Dim index As Integer = 2
                Do While True
                    Dim num9 As Integer = num7
                    If (index > num9) Then
                        If Versioned.IsNumeric(expression) Then
                            num2 = Conversions.ToInteger(expression)
                        End If
                        Exit Do
                    End If
                    expression = (expression & strArray(index))
                    index += 1
                Loop
            End If
            Return num2
        End Function

        Public Shared Function ValidateSecurePropertyKey(ByVal Key As String) As Boolean
            Return Misc.ValidateSecurePropertyKey(Key, False, False)
        End Function

        Public Shared Function ValidateSecurePropertyKey(ByVal Key As String, ByVal MsgExpired As Boolean, ByVal ThrowOnExpired As Boolean) As Boolean
            Dim keyIsValid As Boolean
            If (Key = Nothing) Then
                If (MsgExpired Or ThrowOnExpired) Then
                    Interaction.MsgBox("Key not provided, and thus is not valid.", MsgBoxStyle.Critical, Nothing)
                End If
                keyIsValid = False
            Else
                Dim key As New clsSecurePropertyKey(Key)
                If ((MsgExpired Or ThrowOnExpired) AndAlso Not key.KeyIsValid) Then
                    Dim message As String = ("The security key provided expired on " & Conversions.ToString(key.KeyExpirationDate.Date) & " and is no longer valid")
                    If ThrowOnExpired Then
                        Throw New Exception(message)
                    End If
                    If MsgExpired Then
                        Interaction.MsgBox(message, MsgBoxStyle.Critical, Nothing)
                    End If
                End If
                keyIsValid = key.KeyIsValid
            End If
            Return keyIsValid
        End Function

    End Class
End Namespace

File: PST/FUEL Source Code/FUEL/Backup/FS/Misc.cs
// Decompiled with JetBrains decompiler
// Type: FUEL.FS.Misc
// Assembly: FUEL, Version=2.4.8.0, Culture=neutral, PublicKeyToken=null
// MVID: A3D630D0-0620-425D-8CFB-B67B2AC75507
// Assembly location: C:\hp\FUEL\FUEL.dll

using DataMatrix.net;
using FUEL.My;
using MessagingToolkit.QRCode.Codec;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using System;
using System.Diagnostics;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;
using System.IO;
using System.Reflection;
using System.Text;

namespace FUEL.FS
{
  [StandardModule]
  public sealed class Misc
  {
    public static void CheckForDAQDrivers()
    {
      string str = Path.Combine(new System.IO.FileInfo(Assembly.GetExecutingAssembly().Location).DirectoryName, "LaunchNIDaqInstall.exe");
      Process process = new Process();
      process.StartInfo.WindowStyle = ProcessWindowStyle.Hidden;
      process.StartInfo.FileName = str;
      process.Start();
      process.WaitForExit();
    }

    public static void CheckForUpdate(CheckForUpdates.CheckType UpdateType) => modCheckForUpdates.CheckForUpdate(CheckForUpdates.CheckType.FS);

    public static object GetComputerName() => Misc.GetComputerName(false);

    internal static object GetComputerName(bool ForFUELCompatFileName) => !ForFUELCompatFileName ? (object) MyProject.Computer.Name : (object) MyProject.Computer.Name.Replace("_", "-");

    public static object GenerateFUELCompatibleFileName(
      string Product,
      string DataType,
      ctrlUploadFiles.FileType ValidDuration,
      string FileExtension)
    {
      return Misc.GenerateFUELCompatibleFileName(Product, DataType, ValidDuration, FileExtension, (string) null);
    }

    public static object GenerateFUELCompatibleFileName(
      string Product,
      string DataType,
      ctrlUploadFiles.FileType ValidDuration,
      string FileExtension,
      string ExtraInfo)
    {
      string Right = Strings.Replace(Conversions.ToString(DateAndTime.Now.Date), "/", "-");
      if (ValidDuration == ctrlUploadFiles.FileType.Monthly)
        Right = Strings.Replace(Conversions.ToString(DateAndTime.Now.Month) + "-" + Conversions.ToString(DateAndTime.Now.Year), "/", "-");
      if (!FileExtension.StartsWith("."))
        FileExtension = "." + FileExtension;
      string str = Conversions.ToString(Operators.ConcatenateObject(Operators.ConcatenateObject(Operators.ConcatenateObject((object) (DataType + "_" + Product + "_"), Misc.GetComputerName(true)), (object) "_"), (object) Right));
      if (Operators.CompareString(ExtraInfo, (string) null, false) != 0)
        str = str + "_" + ExtraInfo;
      return (object) (str + FileExtension);
    }

    public static bool ValidateSecurePropertyKey(string Key) => Misc.ValidateSecurePropertyKey(Key, false, false);

    public static bool ValidateSecurePropertyKey(string Key, bool MsgExpired, bool ThrowOnExpired)
    {
      if (Operators.CompareString(Key, (string) null, false) != 0)
      {
        clsSecurePropertyKey securePropertyKey = new clsSecurePropertyKey(Key);
        if (MsgExpired | ThrowOnExpired && !securePropertyKey.KeyIsValid)
        {
          string str = "The security key provided expired on " + Conversions.ToString(securePropertyKey.KeyExpirationDate.Date) + " and is no longer valid";
          if (ThrowOnExpired)
            throw new Exception(str);
          if (MsgExpired)
          {
            int num = (int) Interaction.MsgBox((object) str, MsgBoxStyle.Critical);
          }
        }
        return securePropertyKey.KeyIsValid;
      }
      if (MsgExpired | ThrowOnExpired)
      {
        int num1 = (int) Interaction.MsgBox((object) "Key not provided, and thus is not valid.", MsgBoxStyle.Critical);
      }
      return false;
    }

    public static void CreateQRCode(string PSID, string PrinterSerialNum, string OutputPath) => Misc.CreateQRCode(PSID, PrinterSerialNum, OutputPath, 4, 2, false);

    public static void CreateQRCode(
      string PSID,
      string PrinterSerialNum,
      string OutputPath,
      int Scale,
      int ErrorCorrection,
      bool TransParentBackground)
    {
      FileSystem.SetPath(Path.GetDirectoryName(OutputPath));
      Color color1 = new Color();
      Color color2 = TransParentBackground ? Color.White : Color.Snow;
      Bitmap OriginalImage = new QRCodeEncoder()
      {
        QRCodeErrorCorrect = ((QRCodeEncoder.ERROR_CORRECTION) ErrorCorrection),
        QRCodeScale = Scale,
        QRCodeEncodeMode = QRCodeEncoder.ENCODE_MODE.ALPHA_NUMERIC,
        QRCodeBackgroundColor = color2
      }.Encode(PSID);
      OriginalImage.SetPixel(1, 1, Color.Red);
      Misc.AddTextToBarcode(PSID + "\r\n" + PrinterSerialNum + "\r\n" + Conversions.ToString(DateAndTime.Now) + "\r\nFUEL: " + PST.TestInformation.FuelRev, OriginalImage, TransParentBackground).Save(OutputPath, ImageFormat.Bmp);
    }

    public static void Create2dBarcode(
      string PSID,
      string PrinterSerialNum,
      string OutputPath,
      bool TransParentBackground)
    {
      FileSystem.SetPath(Path.GetDirectoryName(OutputPath));
      Color color1 = new Color();
      Color color2 = TransParentBackground ? Color.White : Color.Snow;
      DmtxImageEncoder dmtxImageEncoder = new DmtxImageEncoder();
      Misc.AddTextToBarcode(PSID + "\r\n" + PrinterSerialNum + "\r\n" + Conversions.ToString(DateAndTime.Now) + "\r\nFUEL: " + PST.TestInformation.FuelRev, dmtxImageEncoder.EncodeImage(PSID, new DmtxImageEncoderOptions()
      {
        ModuleSize = 9,
        BackColor = color2,
        MarginSize = 0,
        ForeColor = Color.Black
      }), TransParentBackground).Save(OutputPath, ImageFormat.Bmp);
    }

    public static void CreateBMPFromString(
      string Text,
      string OutputPath,
      bool TransParentBackground)
    {
      Brush brush1 = TransParentBackground ? Brushes.White : Brushes.Snow;
      Font font1 = new Font("Arial", 20f);
      SizeF sizeF = Graphics.FromImage((Image) new Bitmap(1, 1, PixelFormat.Format24bppRgb)).MeasureString(Text, font1);
      Bitmap bitmap = new Bitmap(checked ((int) System.Math.Round((double) sizeF.Width)), checked ((int) System.Math.Round((double) sizeF.Height)), PixelFormat.Format24bppRgb);
      Graphics graphics1 = Graphics.FromImage((Image) bitmap);
      using (graphics1)
      {
        Graphics graphics2 = graphics1;
        Brush brush2 = brush1;
        Rectangle rectangle = new Rectangle(-1, -1, checked ((int) System.Math.Round((double) unchecked (sizeF.Width + 2f))), checked ((int) System.Math.Round((double) unchecked (sizeF.Height + 2f))));
        Rectangle rect = rectangle;
        graphics2.FillRectangle(brush2, rect);
        Graphics graphics3 = graphics1;
        string s = Text;
        Font font2 = font1;
        Brush red = Brushes.Red;
        rectangle = new Rectangle(1, 1, checked ((int) System.Math.Round((double) unchecked (sizeF.Width + 2f))), checked ((int) System.Math.Round((double) unchecked (sizeF.Height + 2f))));
        RectangleF layoutRectangle = (RectangleF) rectangle;
        graphics3.DrawString(s, font2, red, layoutRectangle);
      }
      bitmap.Save(OutputPath);
    }

    private static Bitmap AddTextToBarcode(
      string bcdText,
      Bitmap OriginalImage,
      bool TransParentBackground)
    {
      Brush brush1 = TransParentBackground ? Brushes.White : Brushes.Snow;
      Font font1 = new Font("Courier New", 24f);
      string text = bcdText;
      SizeF sizeF = Graphics.FromImage((Image) OriginalImage).MeasureString(text, font1);
      Bitmap barcode = new Bitmap(checked ((int) System.Math.Round(unchecked ((double) OriginalImage.Width + (double) sizeF.Width + 12.0))), checked (OriginalImage.Height + 12), PixelFormat.Format24bppRgb);
      Graphics graphics1 = Graphics.FromImage((Image) barcode);
      using (graphics1)
      {
        graphics1.SmoothingMode = SmoothingMode.AntiAlias;
        Graphics graphics2 = graphics1;
        Brush brush2 = brush1;
        RectangleF rectangleF = new RectangleF(-1f, -1f, (float) checked (barcode.Width + 2), (float) checked (barcode.Height + 2));
        RectangleF rect = rectangleF;
        graphics2.FillRectangle(brush2, rect);
        graphics1.DrawImageUnscaled((Image) OriginalImage, new Rectangle(6, 6, barcode.Width, barcode.Height));
        Graphics graphics3 = graphics1;
        string s = text;
        Font font2 = font1;
        Brush red = Brushes.Red;
        rectangleF = new RectangleF((float) checked (OriginalImage.Width + 10), 0.0f, (float) barcode.Width, (float) barcode.Height);
        RectangleF layoutRectangle = rectangleF;
        graphics3.DrawString(s, font2, red, layoutRectangle);
      }
      return barcode;
    }

    public static int GetMediaSize(string File)
    {
      FileSystem.SetPath(File, FilePurpose.ForReading);
      byte[] numArray = MyProject.Computer.FileSystem.ReadAllBytes(File);
      bool flag1 = false;
      string[] arySrc = new string[1];
      long num1 = (long) Information.UBound((Array) numArray);
      long index1 = 0;
      while (index1 <= num1)
      {
        if (numArray[checked ((int) index1)] == (byte) 38 && numArray[checked ((int) (index1 + 1L))] == (byte) 108)
        {
          bool flag2 = false;
          long index2 = index1;
          while (!flag2 & index2 < (long) Information.UBound((Array) numArray))
          {
            if (Operators.CompareString(arySrc[0], (string) null, false) == 0)
            {
              arySrc[0] = Misc.ConvertByteArrayToString(new byte[1]
              {
                numArray[checked ((int) index2)]
              });
            }
            else
            {
              arySrc = (string[]) Utils.CopyArray((Array) arySrc, (Array) new string[checked (Information.UBound((Array) arySrc) + 1 + 1)]);
              arySrc[Information.UBound((Array) arySrc)] = Misc.ConvertByteArrayToString(new byte[1]
              {
                numArray[checked ((int) index2)]
              });
            }
            if (Operators.CompareString(Misc.ConvertByteArrayToString(new byte[1]
            {
              numArray[checked ((int) index2)]
            }).ToLower(), "a", false) == 0)
            {
              flag1 = true;
              goto label_14;
            }
            else
            {
              if (Operators.CompareString(Misc.ConvertByteArrayToString(new byte[1]
              {
                numArray[checked ((int) index2)]
              }), "\u001B", false) == 0)
              {
                flag2 = true;
                flag1 = false;
                arySrc = new string[1];
              }
              checked { ++index2; }
            }
          }
        }
        checked { ++index1; }
      }
label_14:
      int mediaSize = -100;
      if (flag1)
      {
        string Expression = (string) null;
        int num2 = checked (Information.UBound((Array) arySrc) - 1);
        int index3 = 2;
        while (index3 <= num2)
        {
          Expression += arySrc[index3];
          checked { ++index3; }
        }
        if (Versioned.IsNumeric((object) Expression))
          mediaSize = Conversions.ToInteger(Expression);
      }
      return mediaSize;
    }

    private static string ConvertByteArrayToString(byte[] byteArray) => Encoding.UTF8.GetString(byteArray);
  }
}
